yggdrasilffi/src/flat/mod.rs

```diff
@@ -0,0 +1,186 @@
+use flatbuffers::root;
+use std::ffi::c_void;
+
+use messaging::messaging::{
+    ContextMessage, Response,
+};
+use serialisation::{FlatError, FlatMessage, ResponseMessage};
+
+use crate::flat::serialisation::Buf;
+use crate::{get_engine, recover_lock};
+use std::collections::HashMap;
+use unleash_yggdrasil::state::EnrichedContext;
+
+mod serialisation;
+
+#[allow(clippy::all)]
+mod messaging {
+    #![allow(dead_code)]
+    #![allow(non_snake_case)]
+    #![allow(warnings)]
+    include!("enabled-message_generated.rs");
+}
+
+impl TryFrom<ContextMessage<'_>> for EnrichedContext {
+    type Error = FlatError;
+
+    fn try_from(value: ContextMessage) -> Result<Self, Self::Error> {
+        let toggle_name = value.toggle_name().ok_or(FlatError::MissingFlagName)?;
+
+        let context = EnrichedContext {
+            toggle_name: toggle_name.to_string(),
+            runtime_hostname: value.runtime_hostname().map(|f| f.to_string()),
+            user_id: value.user_id().map(|f| f.to_string()),
+            session_id: value.session_id().map(|f| f.to_string()),
+            environment: value.environment().map(|f| f.to_string()),
+            app_name: value.app_name().map(|f| f.to_string()),
+            current_time: value.current_time().map(|f| f.to_string()),
+            remote_address: value.remote_address().map(|f| f.to_string()),
+            properties: value.properties().map(|entries| {
+                entries
+                    .iter()
+                    .filter_map(|entry| Some((entry.key().to_string(), entry.value()?.to_string())))
+                    .collect::<HashMap<String, String>>()
+            }),
+            external_results: value.custom_strategies_results().map(|entries| {
+                entries
+                    .iter()
+                    .map(|entry| (entry.key().to_string(), entry.value()))
+                    .collect::<HashMap<String, bool>>()
+            }),
+        };
+
+        Ok(context)
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn free_buf(buf: Buf) {
+    if buf.ptr.is_null() {
+        return;
+    }
+    unsafe {
+        // We assume capacity == len. If that's not guaranteed, include `cap` in Buf.
+        let _ = Vec::from_raw_parts(buf.ptr, buf.len as usize, buf.len as usize);
+        // Dropping the Vec frees the memory with Rust's allocator.
+    }
+}
+
+#[unsafe(no_mangle)]
+pub extern "C" fn flat_get_state(engine_ptr: u32) -> u32 {
+    let lock = unsafe { get_engine(engine_ptr as *mut c_void).unwrap() };
+    let engine = recover_lock(&lock);
+
+    let state = engine.get_state();
+    if let Ok(json_str) = serde_json::to_string(&state) {
+        use std::ffi::CString;
+        let c_string = CString::new(json_str).unwrap();
```

Comment:
Okay fine, so  this is probably one of those places where unwrap is okay :sweat_smile: